benchmark.cpp

```diff
@@ -3,50 +3,79 @@
 }
 
-template <typename T>
-void benchMarkTrusted(const T *codeUnits)
+template <typename String>
+void benchMarkTrusted(const String &string, int maxCount = BENCH_COUNT)
 {
-    const auto length = std::char_traits<T>::length(codeUnits);
     using clock = std::chrono::steady_clock;
-    auto view = kdab::make_trusted_unicode_view(codeUnits, codeUnits + length);
+    auto view = kdab::make_trusted_unicode_view(string.begin(), string.end());
 
     auto start = clock::now();
-    for (int i = 0; i < BENCH_COUNT; ++i) {
+    for (int i = 0; i < maxCount; ++i) {
         for (auto codePoint : view)
           escape(&codePoint);
     }
     clobber();
     auto stop = clock::now();
-    std::cout << "Benchmarking (trusted) " << length << " code units of utf" << 8 * sizeof(T) << " took: " << std::chrono::duration_cast<std::chrono::milliseconds>(stop - start).count() << "ms " << std::endl;
+    std::cout << "Benchmarking (safe) " << string.size() << " code units of utf" << 8 * sizeof(typename String::value_type) << " took: " << std::chrono::duration_cast<std::chrono::milliseconds>(stop - start).count() << "ms " << std::endl;
+}
+
+enum {
+    BULK_DATA_COPIES = 100000,
+    BULK_DATA_TEST_COUNT = 100
+};
+
+template <typename String>
+String prepareBulkData(const String &input)
+{
+    String result;
+    result.reserve(input.size() * BULK_DATA_COPIES);
+    for (int i = 0; i < BULK_DATA_COPIES; ++i)
+        std::copy(input.begin(), input.end(), std::back_inserter(result));
+
+    return result;
+}
+
+template <typename String>
+void benchmarkSafeBulkData(const String &string)
+{
+    String data = prepareBulkData(string);
+    benchMarkSafe(data, BULK_DATA_TEST_COUNT);
+}
+
+template <typename String>
+void benchmarkTrustedBulkData(const String &string)
+{
+    String data = prepareBulkData(string);
+    benchMarkTrusted(data, BULK_DATA_TEST_COUNT);
 }
 
 int main()
 {
-    const char     str8[]  = u8"€Hello𐐷7バイト以上の文字は規定されないため、0xFE、0xFF (BOM) に0xFEと0xFFを使用するUTF-16やUTF-32が、UTF-8と混同されることはない。UTF-16 is specified in the latest versions of both the international standard ISO/IEC 10646 and the Unicode Standard.";
-    const char16_t str16[] = u"€Hello𐐷7バイト以上の文字は規定されないため、0xFE、0xFF (BOM) に0xFEと0xFFを使用するUTF-16やUTF-32が、UTF-8と混同されることはない。UTF-16 is specified in the latest versions of both the international standard ISO/IEC 10646 and the Unicode Standard.";
-    const char32_t str32[] = U"€Hello𐐷7バイト以上の文字は規定されないため、0xFE、0xFF (BOM) に0xFEと0xFFを使用するUTF-16やUTF-32が、UTF-8と混同されることはない。UTF-16 is specified in the latest versions of both the international standard ISO/IEC 10646 and the Unicode Standard.";
+    std::string str8  = u8"€Hello𐐷7バイト以上の文字は規定されないため、0xFE、0xFF (BOM) に0xFEと0xFFを使用するUTF-16やUTF-32が、UTF-8と混同されることはない。UTF-16 is specified in the latest versions of both the international standard ISO/IEC 10646 and the Unicode Standard.";
```

Comment:
You cheater!
I wondered why your code was faster or the same as mine, now I know why ;-)